### DISCUSSION TITLE
Windows Community Toolkit v8.0.230907

### DISCUSSION THREAD

**Discussion by michael-hawker** [CommunityToolkit/Windows#228] [Announcements]
# _[See More Information in Release Blog Here](https://devblogs.microsoft.com/ifdef-windows/announcing-windows-community-toolkit-v8-0)_

## <a name="changes"></a> New Package Structure & Breaking Changes 🚨

If you've been following our [previous blogs](https://devblogs.microsoft.com/ifdef-windows/windows-community-toolkit-8-0-pre-release/), you know we've made changes to how namespaces and packages work with 8.0.

We've restructured many packages for a few reasons, but the main benefit is it allows developers targeting or migrating between multiple platforms to easily migrate code written on 8.0 without having to change their source. It also enables library developers to multi-target components with XAML easily.

### <a name="namespace"></a> One namespace to rule them all 💍

Regardless of which new features and packages you use ([see below](#features)), the **namespaces** for code within the packages are now **unified** with a root of `CommunityToolkit.WinUI.*`. This ensures that if you target multiple platforms or migrate between them, no code referencing the Toolkit needs to change (just the package)!

### <a name="packagenaming"></a> Package Naming 📦

Previously, starting with 7.x, UWP packages started with `Microsoft.Toolkit.Uwp.UI.*` and had similar namespaces, and WindowsAppSDK package started wih `CommunityToolkit.WinUI.UI.*` and had a similar namespace. [See this blog for more information on our package identity journey](https://devblogs.microsoft.com/ifdef-windows/windows-community-toolkit-for-project-reunion-0-5/#package-identity-and-renaming). These packages will be deprecated and point to their 8.0 equivalents (continue reading).

There are now two sets of packages for each component:

- `CommunityToolkit.Uwp.*` for UWP + WinUI 2 or Uno.UI
- `CommunityToolkit.WinUI.*` for Windows App SDK + WinUI 3 or Uno.WinUI

Regardless of  the package chosen, as mentioned above, all namespaces have converged on `CommunityToolkit.WinUI.*` (_note the dropping of the additional `.UI.`_ in namespace and package names).

If you're using [Uno Platform](https://platform.uno), the **same** package can be used across your various platform heads now! Just be sure to select the package that corresponds to the target of your **Windows** platform head, either `.Uwp.` or `.WinUI.` for the Windows App SDK corresponding to `Uno.UI` and `Uno.WinUI`, respectively.

### <a name="removed"></a> Removed Features 🦴

The Toolkit is huge! It's had a lot of features added over the course of the last **7** years. Unfortunately, this means we couldn't migrate everything in this initial transition. Also, some features may not be available on all platforms due to platform limitations or unimplemented APIs.

Our initial focus has been on **UWP** and **Windows App SDK** compatibility, followed by **Web Assembly (WASM)**. If you find an issue on a particular platform, [please let us know](https://github.com/CommunityToolkit/Windows/discussions), but we may defer you to the underlying platform's repository if an API is missing.

If there's a feature from the Toolkit that wasn't migrated (and **not** listed below), [open a discussion](https://github.com/CommunityToolkit/Windows/discussions) to garner support for future migration or to get information on contributing to migrate it for everyone.

The following is a partial list of _known things_ which are **not** included in this release:

- AdaptiveGridView - [See UniformGridLayout for ItemsRepeater](https://learn.microsoft.com/windows/apps/design/controls/items-repeater#uniformgridlayout).
- DataGrid - DataGrid is a project of its own with over 30k lines of code... In the meantime, check out the [Windows Community Toolkit Labs](https://aka.ms/toolkit/labs/windows) [DataTable](https://github.com/CommunityToolkit/Labs-Windows/discussions/415) experiment.
- DropShadowPanel - This has been replaced by our own Attached Shadows in the `Extensions` and `Media` package.
- [Expander](https://learn.microsoft.com/windows/apps/design/controls/expander) - This is part of WinUI now.
- InAppNotification - This control has been replaced by the `StackedNotificationsBehavior` in our `Behaviors` package to build on top of the platform `InfoBar` control.
- MarkdownTextBlock - This control had a custom Markdown Parser, we didn't want to port the existing codebase because of this. However, the community is working on a replacement in a [WCT Labs PR here](https://github.com/CommunityToolkit/Labs-Windows/pull/480) based on the excellent [Markdig](https://github.com/xoofx/markdig) library.
- [Menu](https://learn.microsoft.com/windows/apps/design/controls/menus#create-a-menu-bar) - This is part of WinUI now, see `MenuBar`.
- Notifications - The notification helpers in the Toolkit have migrated to the [Windows App SDK](https://aka.ms/windowsappsdk), see their [documentation here](https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/notifications/app-notifications/) to get started. [Direct issues to the WindowsAppSDK.](https://aka.ms/windowsappsdk)
- [RadialGradientBrush](https://learn.microsoft.com/windows/winui/api/microsoft.ui.xaml.media.radialgradientbrush) - This is part of WinUI now.
- RadialProgressBar - This is part of WinUI now, see [ProgressRing documentation and set IsIndeterminate to false](https://learn.microsoft.com/windows/apps/design/controls/progress-controls#progressring---determinate).
- [XAML Islands Helpers](https://github.com/CommunityToolkit/Microsoft.Toolkit.Win32) - The tooling in this repository is no longer maintained. XAML Islands is part of the [Windows App SDK](https://aka.ms/windowsappsdk) and [shipped in v1.4](https://learn.microsoft.com/windows/apps/windows-app-sdk/stable-channel#xaml-islands-no-longer-experimental). [Direct issues to the WindowsAppSDK.](https://aka.ms/windowsappsdk)

### <a name="migration"></a> Migration Notes 🚚

_Be sure to read the rest of the breaking change notes above!_ 😉

Most UI controls have their own package now under `CommunityToolkit.<Uwp|WinUI>.Controls.*`; we're working to have a meta package again, but _do not have one at this time_.

For individual grouped control packages of 7.x, you can find the following changes:

- From Controls.Core:
  - `DropShadowPanel` has been replaced with **Attached Shadows** in the `Extensions` and `Media` packages.
  - `InAppNotification` is now the `StackedNotificationsBehavior` for WinUI's `InfoBar` in the `Behaviors` package.
  - `Menu` is part of WinUI now, see `MenuBar`.
  - `RadialProgressBar` is part of WinUI's `ProgressRing` now, IsIndeterminate="False".
- From Controls.Layout:
  - `Expander` is part of WinUI now.
  - `GridSplitter` is part of the `Sizers` package, it's also been entirely re-written, but with defaults should look/behave the same, if not better!
  - "Headered" controls are part of the `HeaderedControls` package, `HeaderedItemsControl` now uses `ItemsPresenter`'s `Header` and `Footer` **new**, `Orientation` has been removed, customize the `ItemsPanel` instead.
- For parts of the `Uwp/WinUI` and `Uwp/WinUI.UI` packages, look for our `Collections`, `Converters`, `Extensions`, `Helpers`, and `Triggers` packages.

Other controls either have their own package or have not been migrated yet. [Search and then open a discussion](https://github.com/CommunityToolkit/Windows/discussions) if you need assistance.

## <a name="packages"></a> Packages/Feature Overview 📦

Search for either `CommunityToolkit.Uwp` or `CommunityToolkit.WinUI` (Windows App SDK), depending on your target platform, and one of the packages listed below should appear. Today's release is version `8.0.230907`.

- **Animations** - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Animations) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Animations)
- **Behaviors** ✨ - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Behaviors) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Behaviors) ✨
- **Collections** - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Collections) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Collections)
- Controls - All the following packages are under the `CommunityToolkit.WinUI.Controls` namespace (see above):
  - **CameraPreview** 💻 - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Controls.CameraPreview) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Controls.CameraPreview)
  - **HeaderedControls** ✨ - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Controls.HeaderedControls) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Controls.HeaderedControls) ✨
  - **ImageCropper** 💻 - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Controls.ImageCropper) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Controls.ImageCropper)
  - **LayoutTransformControl** - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Controls.LayoutTransformControl) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Controls.LayoutTransformControl)
  - **MetadataControl** - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Controls.MetadataControl) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Controls.MetadataControl)
  - **Primitives** - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Controls.Primitives) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Controls.Primitives)
  - **RadialGauge** 💻🎨✨ - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Controls.RadialGauge) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Controls.RadialGauge) ✨
  - **RangeSelector** 🎨 - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Controls.RangeSelector) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Controls.RangeSelector)
  - **RichSuggestBox** 💻🎨 - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Controls.RichSuggestBox) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Controls.RichSuggestBox)
  - **Segmented** ✨ - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Controls.Segmented) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Controls.Segmented) ✨
  - **SettingsControls** ✨ - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Controls.SettingsControls) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Controls.SettingsControls) ✨
  - **Sizers** 🎨✨ - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Controls.Sizers) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Controls.Sizers) ✨
  - **TokenizingTextBox** 🎨 - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Controls.TokenizingTextBox) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Controls.TokenizingTextBox)
- **Converters** - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Converters) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Converters)
- **DeveloperTools** - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.DeveloperTools) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.DeveloperTools)
- **Extensions** - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Extensions) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Extensions)
- **Helpers** - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Helpers) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Helpers)
- **Lottie** 💻 - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Lottie) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Lottie)
- **Media** 💻 - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Media) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Media)
- **Triggers** - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Triggers) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Triggers)

'✨' denotes new or updated packages/features from version 7.1.

'🎨' denotes updated XAML styling from version 7.1.

'💻' denotes packages which only target _UWP/WindowsAppSDK_. However, its absence does not imply that every feature will work across all platforms (see note in introduction).

🔥🐉 **Note**: _Mixing and matching_ these new 8.0 packages with existing Windows Community Toolkit 7.x packages, especially UI controls, can lead to issues with the namespace changes and dependencies on common code (like the Visual Tree extensions). If there's something that we haven't yet ported but you need, [open a discussion](https://github.com/CommunityToolkit/Windows/discussions) to let us know or claim it and contribute!

## What's Changed since pre-releases

* Update tooling submodule for rebrand changes by @michael-hawker in https://github.com/CommunityToolkit/Windows/pull/189
* [Docs] Minor tweaks by @niels9001 in https://github.com/CommunityToolkit/Windows/pull/198
* Porting the Connected Animations sample by @niels9001 in https://github.com/CommunityToolkit/Windows/pull/180
* Snap to Uno-check 1.13 manifest to resolve build issue by @michael-hawker in https://github.com/CommunityToolkit/Windows/pull/202
* Update tooling for app changes by @michael-hawker in https://github.com/CommunityToolkit/Windows/pull/200
* Include component icons as library assets by @Arlodotexe in https://github.com/CommunityToolkit/Windows/pull/203
* Sample App Package/Source Links by @michael-hawker in https://github.com/CommunityToolkit/Windows/pull/204
* Add source syntax highlighting to sample app by @Arlodotexe in https://github.com/CommunityToolkit/Windows/pull/205
* Niels9001/recategorization by @niels9001 in https://github.com/CommunityToolkit/Windows/pull/211
* Minor sample tweaks by @niels9001 in https://github.com/CommunityToolkit/Windows/pull/209
* Removed all markdown links starting with /dotnet/api/ by @Arlodotexe in https://github.com/CommunityToolkit/Windows/pull/213
* Sample App Final Build by @michael-hawker in https://github.com/CommunityToolkit/Windows/pull/214
* Do a pass on spelling and markdown linting in markdown only by @michael-hawker in https://github.com/CommunityToolkit/Windows/pull/216
* [SettingsExpander] Fix for CornerRadius by @niels9001 in https://github.com/CommunityToolkit/Windows/pull/218
* Update Build so that Uwp package doesn't include WASDK and vice-versa by @michael-hawker in https://github.com/CommunityToolkit/Windows/pull/221
* Update ReadMe.md for release by @michael-hawker in https://github.com/CommunityToolkit/Windows/pull/225
* Updating Collections namespace by @niels9001 in https://github.com/CommunityToolkit/Windows/pull/226
* Add Readmes to each component by @michael-hawker in https://github.com/CommunityToolkit/Windows/pull/224

**Full Changelog**: https://github.com/CommunityToolkit/Windows/compare/v8.0.230828-rc...v8.0.230907

<hr /><em>This discussion was created from the release <a href='https://github.com/CommunityToolkit/Windows/releases/tag/v8.0.230907'>v8.0.230907</a>.</em>

**Comment by DominicMaas** [comment]
Just upgraded, pretty smooth! Thanks for the release. Something I noticed during migration, there are two `IIncrementalSource` interfaces.

`CommunityToolkit.Common.Collections.IIncrementalSource<T>`

and


`CommunityToolkit.WinUI.Collections.IIncrementalSource<T>`

They appear to be identical (duplicate file?). Only `CommunityToolkit.WinUI.Collections.IIncrementalSource<T>` can be used with `IncrementalLoadingCollection`.

**Reply by michael-hawker** [reply]
Thanks @DominicMaas this may have been a leftover oversight in the migration between the .NET Community Toolkit spinning out of the Windows Community Toolkit... 🤦‍♂️

@Sergio0694 do you remember? Were we supposed to update the Windows one to use the .NET one and forgot when we ported?

**Comment by michael-hawker** [comment]
There's a known issue with the sample app not displaying XAML/C# code correctly for some reason, we think it's a build authoring issue, but not sure, we're still investigating. Please bear with us!

**Reply by Arlodotexe** [reply]
Following this here - https://github.com/CommunityToolkit/Windows/issues/230

**Reply by michael-hawker** [reply]
This is fixed in the store app in Sample App Version 8.0.2. If you have this version and still see problems, let us know.

**Comment by tipa** [comment]
Happy to see this project is still alive!
Would love to upgrade, but it looks most of the parts I was using were removed, most importantly `AdaptiveGridView` & the Notifications helper.
I am using UWP so it doesn't help me if the Notifications helper is now in the WindowsAppSDK. And I am using GridView that the user can interact with (click, multi-select, etc), so `ItemsRepeater` + `UniformGridLayout` isn't an alternative to `AdaptiveGridView`. Is there any hope that those bits will make it back into the toolkit in a future update?

**Comment by 4OTbIPE** [comment]
I can't find the SystemInformation class in the 8.0 packages from Microsoft.Toolkit.Uwp.Helpers. Is it not available in the new version? There were convenient properties IsFirstRun, IsAppUpdated, etc. I also can't find Print Helper. It's not even in the Toolkit Gallery app(

**Reply by 4OTbIPE** [reply]
ImageEx and Carousel were also lost. I didn't see the information about their removal in the description of the deleted components. Is their absence planned, or did they not get into the package by mistake?

**Reply by michael-hawker** [reply]
There's a lot of code in the toolkit, we're also a small team, so there's only so much that we could do for this release and we had to make a lot of hard decisions. There are not many folks that are contributing to maintain many of these components either. We're happy to work with anyone who wants to help update and modernize any components that it still makes sense to have in the Toolkit.

Some like `SystemInformation` require more work for supporting the Windows App SDK with packaged and unpackaged apps, so it's a harder thing to port, but still happy for someone to pick that up if they'd like to see it. It was higher on our list, but required more effort than other components which could just be copied over.

Others like `ImageEx` and `Carousel` are much older and probably should be reworked or rethought more careful. We've had a lot of folks trying to use `ImageEx` for massive galleries, but it wasn't its original intent to be a high-performance control.

**Reply by tipa** [reply]
@michael-hawker I actually checked out the repo and tried to build/run it earlier today. Without luck -  I didn't even find a .sln file to open the solution in Visual Studio. There's also no guidance on how to open the project in the Readme. If you are hoping for the community to help, then it might be worth considering to make onboarding easier.
Now after some more trying (and struggling) I found out that running the `GenerateAllSolution.bat` script (and installing some more `dotnet` dependencies) opens the project in Visual Studio, but I was presented with a ton of errors, not being able to build a single project. Pretty confusing and frustrating...

**Reply by michael-hawker** [reply]
@tipa because of our submodule for shared infrastructure we use `slngen` to generate solutions, that's what the `GenerateAllSolution.bat` (and `OpenSolution.bat in each component`) do.

Building the whole sample app should be working (as we released our sample app to the store), though individual components can have issues #169 that we're still investigating that make them a bit more difficult out-of-the-box.

For your specific issue, if you could share more details on the errors you're seeing (maybe in a new issue?); we're in the process of updating our guidance and wiki to ensure it's easy for anyone to build and contribute to the project. That's why we made simple stand-alone components that can be opened and worked on without having to build the whole toolkit and encapsulate the source, samples, docs, and tests together.

**Comment by adamfierek** [comment]
Where is Carousel control?

**Reply by Arlodotexe** [reply]
Let's move the discussion over to https://github.com/CommunityToolkit/Labs-Windows/discussions/498

**Comment by ttustonic** [comment]
I see that ListDetailsView is not included in this release. Are there plans to include it later, or suggestion/example how to replace it?
Thanks, Tom

**Reply by unliftedq** [reply]
@michael-hawker could you please help clarify it?

**Reply by JoeTomkinson** [reply]
Only just noticed this, I need to make a list of all the things missing and cross-reference it so I know when I can safely migrate some apps across.

Couldn't see reference to this one in the list either.

**Reply by ms162223** [reply]
I too use the ListDetailsView and ListDetailsViewState.  If this has been deprecated, are there any tutorials on how to replace it with something similar?

**Comment by zsolt3991** [comment]
Hello

Looking forward at migration to the new WinUI packages since the 7.1.x was long overdue an update. The legacy packages contained some generic implementations of Settings/File storage, namely: ApplicationDataStorageHelper, which implements a set of interfaces from "CommunityToolkit.Common"

Are there plans to have some built in support going forward? Or should applications roll their own implementation for both Packaged/Unpackaged WinUI apps?

**Reply by michael-hawker** [reply]
The community has been driving an effort around a Storage abstraction in Labs, [find more info here](https://github.com/CommunityToolkit/Labs-Windows/discussions/229).

**Comment by xiaoliangshi** [comment]
For the projects using DataGrid, is it recommended to keeping using the 7.X version until it is available in 8.X version? Will 7.X version keep maintain for any security/reliability bug fixing?

**Reply by michael-hawker** [reply]
Yes, DataGrid is isolated and should continue to work as the 7.x version alongside any other 8.x components. There are no current plans to service the component, our old pipelines for the [old WindowsCommunityToolkit repo](https://github.com/CommunityToolkit/WindowsCommunityToolkit) are non-functional; So, if we needed to, it'd make sense to migrate it to [our new shared infrastructure instead](https://github.com/CommunityToolkit/Tooling-Windows-Submodule). If anyone has interest in owning that sort of effort, please let us know.

Our current plan has been to sync and coordinate with the platform on [their efforts around a TableView component](https://aka.ms/winappsdk/plans). Thanks!
